charms/kfp-api/src/charm.py

```diff
@@ -289,40 +289,48 @@ def _get_interfaces(self):
         return interfaces
 
     def _get_mysql(self):
-        mysql = self.model.relations["mysql"]
-        if len(mysql) == 0:
-            raise CheckFailedError("Missing required relation for mysql", BlockedStatus)
-        elif len(mysql) > 1:
-            raise CheckFailedError("Too many mysql relations", BlockedStatus)
+        """Returns mysql relation data from the relation with a mysql database.
 
-        try:
-            mysql = mysql[0]
-            unit = list(mysql.units)[0]
-            mysql = mysql.data[unit]
-        except Exception as e:
-            self.log.exception(
-                f"Encountered the following exception when parsing mysql relation: " f"{str(e)}"
-            )
-            raise CheckFailedError(
-                "Unexpected error when parsing mysql relation.  See logs", BlockedStatus
-            )
+        Raises:
+            CheckFailed(..., BlockedStatus) if there is no mysql relation
+            CheckFailed(..., WaitingStatus) if The remove unit has not joined the relation
+            CheckFailed(..., WaitingStatus) if the relation data bag is empty
+        """
+        mysql_relation = self.model.get_relation("mysql")
 
-        expected_attributes = ["database", "host", "root_password", "port"]
+        # Raise exception and stop execution if the mysql relation is not established
+        if not mysql_relation:
+            raise CheckFailedError("There is no mysql relation", BlockedStatus)
+
+        if not mysql_relation.units:
+            raise CheckFailedError("Waiting for remote unit to join relation", WaitingStatus)
 
+        if not mysql_relation.data:
+            raise CheckFailedError("There is no data in the mysql relation", WaitingStatus)
+
+        # This charm should only establish a relation with exactly one unit
+        # the following extracts exactly one unit from the set that's
+        # returned by mysql_relation.data
+        for unit in mysql_relation.units:
+            # Get mysql relation data
+            mysql_relation_data = mysql_relation.data[unit]
```

Comment:
```suggestion
        mysql_unit = mysql_relation.units[0]
        mysql_relation_data = mysql_relation.data[mysql_unit]
```

Reply:
I removed this by error, I've added it again, though it's a bit different of what you are proposing. Let me know what you think.